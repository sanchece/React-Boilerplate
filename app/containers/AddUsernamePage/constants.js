/*
 *
 * AddUsernamePage constants
 *
 */

export const UPDATE_USERNAME = 'app/AddUsernamePage/UPDATE_USERNAME';
export const BLANK_USERNAME = 'Cannot be blank';
