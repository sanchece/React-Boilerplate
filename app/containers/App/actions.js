/*
 *
 * MainPage actions
 *
 */

import {
  USERNAMES_REQUESTED,
  USERNAMES_REQUEST_FAILED,
  USERNAMES_RECEIVED,
  USERNAME_ADD,
  USERNAMES_ADD_FAILED,
  USERNAMES_ADD_SUCCESS,
} from './constants';

export function usernamesRequested() {
  return {
    type: USERNAMES_REQUESTED,
  };
}
export function usernamesRequestFailed(error) {
  return {
    type: USERNAMES_REQUEST_FAILED,
    payload: {
      error,
    },
  };
}

export function usernamesReceived(usernameList) {
  return {
    type: USERNAMES_RECEIVED,
    payload: {
      usernameList,
    },
  };
}

export function usernameAdd(username) {
  return {
    type: USERNAME_ADD,
    payload: {
      username,
    },
  };
}
export function usernameAddFailed(error) {
  return {
    type: USERNAMES_ADD_FAILED,
    payload: {
      error,
    },
  };
}
export function usernameAddSuccess(username) {
  return {
    type: USERNAMES_ADD_SUCCESS,
    payload: {
      username,
    },
  };
}
