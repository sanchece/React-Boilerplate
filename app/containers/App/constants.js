/*
 * AppConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

export const USERNAMES_REQUESTED = 'app/usernamesRequested';
export const USERNAMES_REQUEST_FAILED = 'app/usernamesRequestFailed';
export const USERNAMES_RECEIVED = 'app/usernamesReceived';

export const USERNAME_ADD = 'app/usernameAdd';
export const USERNAMES_ADD_FAILED = 'app/usernameAddFailed';
export const USERNAMES_ADD_SUCCESS = 'app/usernameAddSuccess';

export const USERNAMES_REQUESTED_STATUS = 'Requesting usernames from server...';
export const USERNAMES_REQUEST_FAILED_STATUS =
  'Server usernames request failed.';
export const USERNAMES_RECEIVED_STATUS =
  'Server usernames returned successfully.';

export const USERNAME_ADD_STATUS = 'Adding username to server...';
export const USERNAMES_ADD_FAILED_STATUS = 'Failed to add username to server.';
export const USERNAMES_ADD_SUCCESS_STATUS = 'Username added successfully.';
export const INITIAL_STATE = 'INITIAL_STATE';
