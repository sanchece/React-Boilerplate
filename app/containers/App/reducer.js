/*
 *
 * App reducer
 *
 */
import produce from 'immer';
import {
  USERNAMES_REQUESTED,
  USERNAMES_REQUEST_FAILED,
  USERNAMES_RECEIVED,
  USERNAME_ADD,
  USERNAMES_ADD_FAILED,
  USERNAMES_ADD_SUCCESS,
  USERNAMES_REQUESTED_STATUS,
  USERNAMES_REQUEST_FAILED_STATUS,
  USERNAMES_RECEIVED_STATUS,
  USERNAME_ADD_STATUS,
  USERNAMES_ADD_FAILED_STATUS,
  USERNAMES_ADD_SUCCESS_STATUS,
  INITIAL_STATE,
} from './constants';
export const initialState = {
  usernameList: [],
  loading: {
    status: false,
    getRequestStatus: INITIAL_STATE,
    postRequestStatus: INITIAL_STATE,
  },
  error: false,
};

/* eslint-disable default-case, no-param-reassign */
const appReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      // usernames requested
      case USERNAMES_REQUESTED:
        draft.loading.status = true;
        draft.loading.getRequestStatus = USERNAMES_REQUESTED_STATUS;
        draft.loading.postRequestStatus = INITIAL_STATE;
        draft.error = false;

        break;
      case USERNAMES_REQUEST_FAILED:
        draft.loading.getRequestStatus = USERNAMES_REQUEST_FAILED_STATUS;
        draft.error = action.payload.error;
        draft.loading.status = false;
        break;
      case USERNAMES_RECEIVED:
        if (draft.loading.postRequestStatus !== USERNAME_ADD_STATUS) {
          draft.loading.status = false;
        }
        draft.loading.getRequestStatus = USERNAMES_RECEIVED_STATUS;
        draft.usernameList = action.payload.usernameList;
        break;

      // username added
      case USERNAME_ADD:
        draft.loading.status = true;
        draft.loading.postRequestStatus = USERNAME_ADD_STATUS;
        draft.loading.getRequestStatus = INITIAL_STATE;
        draft.error = false;
        break;
      case USERNAMES_ADD_FAILED:
        draft.loading.status = false;
        draft.loading.postRequestStatus = USERNAMES_ADD_FAILED_STATUS;
        draft.error = action.payload.error;
        break;
      case USERNAMES_ADD_SUCCESS:
        if (draft.loading.getRequestStatus !== USERNAMES_REQUESTED_STATUS) {
          draft.loading.status = false;
        }
        draft.usernameList.unshift(action.payload.username);
        draft.loading.postRequestStatus = USERNAMES_ADD_SUCCESS_STATUS;
        break;
    }
  });

export default appReducer;
