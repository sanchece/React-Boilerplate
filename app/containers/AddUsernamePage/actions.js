/*
 *
 * AddUsernamePage actions
 *
 */

import { UPDATE_USERNAME } from './constants';

export default function updateUsername(username) {
  return {
    type: UPDATE_USERNAME,
    payload: {
      username,
    },
  };
}
