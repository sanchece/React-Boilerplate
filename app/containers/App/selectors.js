import { createSelector } from 'reselect';
import { initialState } from './reducer';

/**
 * Direct selector to the mainPage state domain
 */

const selectAppDomain = state => state.app || initialState;

/**
 * Other specific selectors
 */

const makeUsernamesSelector = () =>
  createSelector(
    selectAppDomain,
    appState => appState.usernameList,
  );
const makeLoadingSelector = () =>
  createSelector(
    selectAppDomain,
    appState => appState.loading,
  );
const makeErrorSelector = () =>
  createSelector(
    selectAppDomain,
    appState => appState.error,
  );

export {
  makeUsernamesSelector,
  makeLoadingSelector,
  makeErrorSelector,
  selectAppDomain,
};
