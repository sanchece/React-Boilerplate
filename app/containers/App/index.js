/**
 *
 * App.js
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 *
 */

import React from 'react';
import { Switch, Route } from 'react-router-dom';
import MainPage from 'containers/MainPage/Loadable';
import AddUsernamePage from 'containers/AddUsernamePage/Loadable';
import NotFoundPage from 'containers/NotFoundPage/Loadable';
import styled from 'styled-components';

import GlobalStyle from '../../global-styles';
import Navbar from '../../components/Navbar';

const AppWrapper = styled.div`
  max-width: calc(768px + 16px * 2);
  margin: 0 auto;
  display: flex;
  min-height: 100%;
  padding: 0 20px;
  flex-direction: column;
  text-align: center;
`;

export default function App() {
  return (
    <AppWrapper>
      <Navbar />

      <Switch>
        <Route exact path="/" component={MainPage} />
        <Route exact path="/add" component={AddUsernamePage} />
        <Route component={NotFoundPage} />
      </Switch>
      <GlobalStyle />
    </AppWrapper>
  );
}
