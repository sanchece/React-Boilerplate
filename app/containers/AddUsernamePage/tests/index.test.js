/**
 *
 * Tests for AddUsernamePage
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
// import 'jest-dom/extend-expect'; // add some helpful assertions
import { Provider } from 'react-redux';
import updateUsername from '../actions';
import { usernameAdd, usernameAddFailed } from '../../App/actions';
import { AddUsernamePage, mapDispatchToProps } from '../index';
import configureStore from '../../../configureStore';
import { DEFAULT_LOCALE } from '../../../i18n';
import { BLANK_USERNAME } from '../constants';
describe('<AddUsernamePage />', () => {
  let store;

  beforeAll(() => {
    store = configureStore();
  });

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    const dispatch = jest.fn();
    render(
      <Provider store={store}>
        <IntlProvider locale={DEFAULT_LOCALE}>
          <AddUsernamePage
            username=""
            error={false}
            loading={{
              status: false,
              getRequestStatus: 'INITIAL_STATE',
              postRequestStatus: 'INITIAL_STATE',
            }}
            dispatch={dispatch}
          />
        </IntlProvider>
      </Provider>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  /**
   * Unskip this test to use it
   *
   * @see {@link https://jestjs.io/docs/en/api#testskipname-fn}
   */
  it('Should render and match the snapshot', () => {
    const dispatch = jest.fn();
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale={DEFAULT_LOCALE}>
          <AddUsernamePage
            username=""
            error={false}
            loading={{
              status: false,
              getRequestStatus: 'INITIAL_STATE',
              postRequestStatus: 'INITIAL_STATE',
            }}
            dispatch={dispatch}
          />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });
});

describe('mapDispatchToProps', () => {
  describe('onUpdateUsername', () => {
    it('should be injected', () => {
      const dispatch = jest.fn();
      const result = mapDispatchToProps(dispatch);
      expect(result.onUpdateUsername).toBeDefined();
    });

    it('should dispatch updateUsername when called', () => {
      const dispatch = jest.fn();
      const result = mapDispatchToProps(dispatch);
      const username = 'carlos';
      result.onUpdateUsername({ target: { value: username } });
      expect(dispatch).toHaveBeenCalledWith(updateUsername(username));
    });
  });

  describe('onSubmitForm', () => {
    it('should be injected', () => {
      const dispatch = jest.fn();
      const result = mapDispatchToProps(dispatch);
      expect(result.onSubmitForm).toBeDefined();
    });

    it('should dispatch usernameAdd when called', () => {
      const dispatch = jest.fn();
      const result = mapDispatchToProps(dispatch);
      const username = 'carlos';
      result.onSubmitForm({ target: { username_input: { value: username } } });
      expect(dispatch).toHaveBeenCalledWith(usernameAdd(username));
    });
    it('should dispatch usernameAddFailed when called username is blank', () => {
      const dispatch = jest.fn();
      const result = mapDispatchToProps(dispatch);
      const username = '';
      result.onSubmitForm({ target: { username_input: { value: username } } });
      expect(dispatch).toHaveBeenCalledWith(usernameAddFailed(BLANK_USERNAME));
    });

    it('should preventDefault if called with event', () => {
      const preventDefault = jest.fn();
      const result = mapDispatchToProps(() => {});
      const username = 'mxstbr';
      const evt = {
        preventDefault,
        target: { username_input: { value: username } },
      };
      result.onSubmitForm(evt);
      expect(preventDefault).toHaveBeenCalledWith();
    });
  });
});
