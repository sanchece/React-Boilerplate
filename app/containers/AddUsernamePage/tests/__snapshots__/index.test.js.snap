// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddUsernamePage /> Should render and match the snapshot 1`] = `
<div>
  <h3>
    Add Username Form
  </h3>
  <form>
    <label
      for="username_input"
    >
      <input
        id="username_input"
        placeholder="Insert username here"
        type="text"
      />
    </label>
    <input
      readonly=""
      type="submit"
      value="Add"
    />
  </form>
  <div />
</div>
`;
